### PR TITLE
Add viz boilerplate

### PR DESCRIPTION
**Pull request by mbovel**
As discussed on Tuesday, here is a structure and boilerplate proposition for the final version of the visualization:

- `scripts` contains pythons scripts we use to download, process and generate the data.
  - `requirements.txt` is a [requirements file](https://pip.pypa.io/en/stable/user_guide/#requirements-files). It provides a simple way to install all python packages used by scripts: `pip install -r requirements.txt`.
  - `generate_mentions.py` is the script I showed you that generates the list of mentions.
  - `utils.py` contains all helper functions. It also contains two constants: `CACHE_DIR` and `DATA_DIR` pointing to the `cache` and `data` directories.
  - We will add other scripts to this folder (for example to download the list of famous persons or to extract the list of main sources from Wikipedia).
- `src` contains the javascript sources for the visualization. There is one class per file, plus the `main.js` file.
  - `Sidebar.js` contains all the logic for the control elements in the sidebar. Its constructor takes two arguments: the HTML element in which to display the controls, and an instance of `GraphPlot`. When the user makes changes on the control elements, It directly calls methods like `setTimeRange` on `GraphPlot` to update the plot.
  - `GraphPlot` is responsible for displaying the graph and interacting with the corresponding SVG code. Its constructor receives one argument: the `svg` element in which to draw the graph.
  - `GraphSimulation` is an helper class for `GraphPlot`. It serves as a model for the underlying physical simulation.
  - `main.js` contains the code responsible for creating the needed classes and initializing the visualization.
- `assets/css` contains all CSS files for the website and the visualization.
  - `website-provided.css` is the CSS code provided at the beginning of the course.
  - `website-custom.css` is our custom code for the website.
  - `visualization.css` is the styles for the general layout of the visualization (mainly responsible for splitting the page in two columns). Specific code for each code is inside the following files.
  - `visualization-sidebar.css` contains the styles for the sidebar.
  - `visualization-graph.css` contains the styles for the graph. 
- `visualization.html` is the entrypoint for the visualization: it includes all needed files and provides a very basic markup.

What do you think about this structure?

### CODE REVIEW

File: src/GraphPlot.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
+class GraphPlot {
+	constructor(/**HTMLElement*/ container) {
+		this.container = d3.select(container);
+		this.simulation = new GraphSimulation();
```
Comment (Aydarkhan) [inline src/GraphPlot.js]: As we discussed last time, some callback functions/class methods in D3 don't allow access to the class itself. That's why I used a global GraphSimulation object. Have you found a workaround?

Comment (Aydarkhan) [inline src/GraphPlot.js]: As we discussed last time, sometimes callback function in D3 which are class methods won't allow access to the class itself, that's why I used a glocal GraphSimulation object. Have you found a workaround?

File: scripts/generate_mentions.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+ORDER BY
+  mentions_count DESC,
+  person,
+  source_name
+LIMIT
+  5000
+""")
+
+output_file = os.path.join(DATA_DIR, 'mentions.csv')
+mentions.to_csv(output_file, index=False)
```
Comment (Aydarkhan) [inline scripts/generate_mentions.py]: To make it cleaner and easier to manipulate with the code, I would make a variable for each query (or a dictionary):
```
queries = {}
queries['mentions'] = f"""....."""
```
Then call it with a single-line code:
`mentions = run_bigquery(name='mentions', sql=queries['mentions'])`

With the dictionary you can have a clean for-loop:
```
for name, sql in queries.items():
  result = run_bigquery(name=name, sql=sql)
  output_file = os.path.join(DATA_DIR, name + '.csv')
  result.to_csv(output_file, index=False)
```

File: visualization.html [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+			integrity="sha256-l85OmPOjvil/SOvVt3HnSSjzF1TUMyT9eV0c2BzEGzU="
+			crossorigin="anonymous"
+		/>
+		<link rel="stylesheet" href="assets/css/visualization.css" />
+		<link rel="stylesheet" href="assets/css/visualization-graph.css" />
+		<link rel="stylesheet" href="assets/css/visualization-sidebar.css" />
+	</head>
+	<body>
+		<main>
+			<svg id="plot" viewBox="-250 -250 500 500" preserveAspectRatio="xMidYMid meet"></svg>
```
Comment (Aydarkhan) [inline visualization.html]: Why -250 ? We could use any limits with d3-force library, but it is easier to grasp something like [0, 500] or [0,100]

File: src/main.js [hunk 4]
```diff
@@ -0,0 +1,4 @@
+const graphPlot = new GraphPlot(document.getElementById("plot"));
+const sidebar = new Sidebar(document.getElementById("sidebar"), graphPlot);
+
+graphPlot.init();
```
Comment (Aydarkhan) [inline src/main.js]: Why using init()? I don't know Javascript well but we already have a _constructor()_ and _init()_ is not a default method.

File: src/GraphPlot.js (hunk 1 above)
Comment (Aydarkhan) [inline src/GraphPlot.js]: As we discussed last time, sometimes D3 callback functions which are class methods won't allow access to the class object itself. That's why I used a global GraphSimulation object. Have you found a workaround?

File: package.json [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/mbovel/dataviz-project.git"
+  },
+  "author": "",
+  "bugs": {
+    "url": "https://github.com/mbovel/dataviz-project/issues"
+  },
+  "homepage": "https://github.com/mbovel/dataviz-project#readme"
+}
```
Comment (Aydarkhan) [inline package.json]: I am just curious where is this file read?

File: src/GraphPlot.js (hunk 1 above)
Comment (mbovel) [reply src/GraphPlot.js]: If I understand the problem well, there here are two potential solutions:

```html
<button id="my-super-button">MySuperButton</button>
<button id="my-super-button-2">MySuperButton2</button>
<body>
<script>
class MySuperButton {
    constructor(/**HTMLElement*/ buttonEl) {
        this.message = 'Hello from MySuperButton!'
        buttonEl.addEventListener('click', this.clickHandler.bind(this))
    }

    clickHandler() {
        alert(this.message)
    }
}

class MySuperButton2 {
    constructor(/**HTMLElement*/ buttonEl) {
        this.message = 'Hello from MySuperButton2!'

        buttonEl.addEventListener('click', () => {
            alert(this.message)
        })
    }
}

new MySuperButton(document.getElementById('my-super-button'))
new MySuperButton2(document.getElementById('my-super-button-2'))
</script>
```

File: scripts/generate_mentions.py (hunk 2 above)
Comment (mbovel) [reply scripts/generate_mentions.py]: > To make it cleaner and easier to manipulate with the code, I would make a variable for each query (or a dictionary).

Not sure this is much cleaner or easier to use, but ok.

> With the dictionary you can have a clean for-loop.

I don't think that would work. To construct to third query, I use results of the previous ones. Also, only the results of the third query (for now) need to be written to a CSV file.

File: visualization.html (hunk 3 above)
Comment (mbovel) [reply visualization.html]: Having the origin at (0, 0) makes more sense to me.

Mike Bostock himself does so in [his example for radial forces](https://bl.ocks.org/mbostock/cd98bf52e9067e26945edd95e8cf6ef9).

Also it enables a simpler definition of the force:

```js
const simulation = d3.forceSimulation(nodesData)
            .force("r", d3.forceRadial(d => d.type === "person" ? 100 : 200))
```

as opposed to this:

```js
const simulation = d3.forceSimulation(nodesData)
            .force("r", d3.forceRadial(d => d.type === "person" ? 100 : 200), height / 2, width / 2)
```

File: src/main.js (hunk 4 above)
Comment (mbovel) [reply src/main.js]: I personally like to separate the code for creating the class from the code to start the animation. I think a user of the class might want to create it and start the animation at a later point. But this is very debatable.

[Folks on SO](https://softwareengineering.stackexchange.com/a/334994) seems to agree with you, so let's follow that and remove the `init` method 😉

File: package.json (hunk 5 above)
Comment (mbovel) [reply package.json]: This is [a file used by npm](https://docs.npmjs.com/files/package.json), the package manager for javascript code.

Here, I only used it to define a `npm run format` command that we can use to automatically clean the code, and potentially set up to work with <https://travis-ci.org/>.

File: src/main.js (hunk 4 above)
Comment (Aydarkhan) [reply src/main.js]: I agree that we start animation at laster point. Moreover, I think that animation should start automatically after we supply the data. And you have already put the `setTimeRange` function. So `init` is not necessary.

File: scripts/generate_mentions.py (hunk 2 above)
Comment (Aydarkhan) [reply scripts/generate_mentions.py]: My bad, if we write to the file only the last one, then we don't need the for-loop and the dictionary. But having SQL query in a separate variable is still a cleaner way in my opinion.

**Pull request merged**